tensorflow/lite/kernels/sub_test.cc

```diff
@@ -72,29 +72,44 @@ class Int64SubOpModel : public BaseSubOpModel {
 
 class QuantizedSubOpModel : public BaseSubOpModel {
  public:
-  using BaseSubOpModel::BaseSubOpModel;
+  QuantizedSubOpModel(TensorData input1, TensorData input2, TensorData output,
+                      ActivationFunctionType activation_type)
+      : BaseSubOpModel(SymmetricInt16Scaling(std::move(input1)),
+                       SymmetricInt16Scaling(std::move(input2)),
+                       SymmetricInt16Scaling(std::move(output)),
+                       activation_type) {}
 
   template <typename integer_dtype>
   std::vector<float> GetDequantizedOutput() {
     return Dequantize<integer_dtype>(ExtractVector<integer_dtype>(output_),
                                      GetScale(output_), GetZeroPoint(output_));
   }
 
-  std::vector<float> GetDequantizedOutputInt16() {
-    return Dequantize<int16_t>(ExtractVector<int16_t>(output_),
-                               GetScale(output_), GetZeroPoint(output_));
+ private:
+  TensorData SymmetricInt16Scaling(TensorData tensor) {
+    // Symmetric range and null zero-point is required for INT16 tensors. As
+    // SingleOpModel::QuantizationParams calculates the scale on an asymmetric
+    // base [int_type::min, int_type::max], manually calculate the scale on a
+    // symmetric range [int_type::min+1, int_type::max] to ensure a null
+    // zero-point.
+    if (tensor.type == TensorType_INT16) {
+      CHECK_EQ(std::abs(tensor.min), tensor.max);
+      tensor.scale = tensor.max / std::numeric_limits<int16_t>::max();
+      tensor.zero_point = 0;
+      tensor.min = 0;
+      tensor.max = 0;
+    }
+
+    return tensor;
   }
 };
 
 // for quantized Sub, the error shouldn't exceed step
-float GetTolerance(int min, int max) {
-  float kQuantizedStep = (max - min) / 255.0;
-  return kQuantizedStep;
-}
-
-float GetToleranceInt16(float min, float max) {
-  float kQuantizedStep = (max - min) / std::numeric_limits<int16_t>::max();
-  return kQuantizedStep;
+template <typename T>
+float GetTolerance(float min, float max) {
+  float kQuantizedStep = (max - min) / (std::numeric_limits<T>::max() -
+                                        std::numeric_limits<T>::min());
+  return 2.0 * kQuantizedStep;
```

Comment:
Where does the 2.0 come from ?

Reply:
Hi @fredrec Thanks for the review.

Tolerance should be 2 quantized steps for the result:
a_float = a_int16 +- quantized_step
if i do difference, then upper bound is 2*quantized_step.

Reply:
@fredrec Can you please take a look on above comments from @wwwind. Thanks!